redis/src/sentinel.rs

```diff
@@ -295,25 +292,80 @@ fn valid_addrs<'a>(
         })
 }
 
-fn check_role_result(result: &RedisResult<Vec<Value>>, target_role: &str) -> bool {
-    if let Ok(values) = result {
-        if !values.is_empty() {
-            if let Ok(role) = String::from_redis_value(&values[0]) {
-                return role.to_ascii_lowercase() == target_role;
-            }
-        }
+fn determine_master_from_role_or_info_replication(connection_info: &ConnectionInfo) -> RedisResult<bool> {
+    let client = Client::open(connection_info.clone())?;
+    let mut conn = client.get_connection()?;
+
+    //Once the client discovered the address of the master instance, it should attempt a connection with the master, and call the ROLE command in order to verify the role of the instance is actually a master.
+    let role = check_role(&mut conn);
+    if role.is_ok_and(|x| matches!(x, Role::Primary { .. })) {
+        return Ok(true);
+    }
+
+    //If the ROLE commands is not available (it was introduced in Redis 2.8.12), a client may resort to the INFO replication command parsing the role: field of the output.
+    let role = check_info_replication(&mut conn);
+    if role.is_ok_and(|x| x == "master") {
+        return Ok(true);
+    }
+
+    //TODO: Maybe there should be some kind of error message if both role checks fail due to ACL permissions?
+    Ok(false)
+}
+
+fn determine_slave_from_role_or_info_replication(connection_info: &ConnectionInfo) -> RedisResult<bool> {
+    let client = Client::open(connection_info.clone())?;
+    let mut conn = client.get_connection()?;
+
+    //Once the client discovered the address of the master instance, it should attempt a connection with the master, and call the ROLE command in order to verify the role of the instance is actually a master.
+    let role = check_role(&mut conn);
+    if role.is_ok_and( |x| matches!(x, Role::Replica { .. })) {
+        return Ok(true);
     }
-    false
+
+    //If the ROLE commands is not available (it was introduced in Redis 2.8.12), a client may resort to the INFO replication command parsing the role: field of the output.
+    let role = check_info_replication(&mut conn);
+    if role.is_ok_and(|x|  x == "slave") {
+        return Ok(true);
+    }
+
+    //TODO: Maybe there should be some kind of error message if both role checks fail due to ACL permissions?
+    Ok(false)
+}
+
+fn old_check_role(connection_info: &ConnectionInfo) -> RedisResult<Role> {
+    let client = Client::open(connection_info.clone())?;
+    let mut conn = client.get_connection()?;
+    let role: RedisResult<Role> = crate::cmd("ROLE").query(&mut conn);
+    return role
+}
+
+fn check_role(conn: &mut Connection) -> RedisResult<Role> {
+    let role: RedisResult<Role> = crate::cmd("ROLE").query(conn);
+    role
```

Comment:
I think that `crate::cmd("ROLE").query(conn)` is sufficient for the full function, no need to set the variable